### ISSUE TITLE
No way to support build flag like `-dynamic-too` (Shake considers change in file pattern grounds for rebuild)

### ISSUE THREAD

**Issue by ezyang** [ndmitchell/shake#382]
GHC supports a flag named `-dynamic-too` with the following semantics:

1. Without `-dynamic-too`, GHC outputs an `.hi` file and an `.o` file when you compile an `.hs` file.
2. With `-dynamic-too`, GHC outputs the same as (1) (bit-for-bit, in fact), AS WELL AS `.dyn_hi` and `.dyn_o`.

I attempted to implement this in Shake by writing a conditional on the `-dynamic-too` flag, providing a `&%>` rule which matched against just `hi` and `o` without the flag, and `hi`, `o`, `dyn_hi` and `dyn_o` with the flag.

However, after implementing this, I ran into mysterious spurious recompilation problems.

The problem is that Shake treats a `&%>` rule as defining a "multiple file rule", with the other rules depending on it. So, without `-dynamic-too`, Shake generates a rule named "*.hi *.o", and with it, it generates a rule "*.hi *.o *.dyn_hi *.dyn_o". If I go from passing `-dynamic-too` to not passing it, the "*.hi *.o" rule doesn't exist so Shake rebuilds it.

@ndmitchell , you've mentioned that you're planning on changing how multi-file rules are implemented. I don't know if this solves this problem. Within the current framework, I don't really know how to solve this.

**Comment by ezyang** [comment]
Using Shake's internal API, I implemented a working version of `-dynamic-too`: https://github.com/ezyang/ghc-shake/blob/master/GhcShake.hs#L670

It's hella janky, but here's the basic idea: I define a new rule key which essentially is a 'Module' (plus some extra info). From this module, I can compute all of the output file paths that will show up if I build it: if I'm building module A, then normally I will get A.hi and A.o, unless `-dynamic-too` is on, then I'll also get A.dyn_hi and A.dyn_o.

The answer type consists of a `Maybe FileA` for every possible output file we can output. `storedValue` loads up `FileA` in much the same way `Files` is implement, but the equality test does something fun: depending on what mode we're in (`-dynamic-too` or not), it will check more or less of the data structure in question. Thus, without `-dynamic-too`, an out-of-date `dyn_hi` file will not cause the validity check to fail.

There is one last important thing, which is what `FileA` gets synthesized when you finish running a rule. I assume that IF you rebuilt an hi/o file, THEN you must rebuild the dyn_hi/dyn_o files next time you run with `-dynamic-too`. (One could imagine a more sophisticated test where if the `hi`/`o` files were unchanging then we could assume the `dyn_hi` and `dyn_o` files were unchanging, but we can't assume this in general.)

Implementing a new rule here was very useful, and it simplified a LOT of code in the build system. It's definitely a keeper.

The recent `storedValueE` patch in #389 doesn't help solve this problem at all: you still need to pick an appropriate answer type for your combined hi/o/dyn_hi/dyn_o rule, and the only sensible answer type involves actually looking at the modtimes/hashes of the files in question

**Comment by ndmitchell** [comment]
While storedValueE might not be useful here, I think it is in general. I haven't figured out a way to capture your dependency pattern without a rule yet.

**Comment by Mathnerd314** [comment]
The dependency graph in your rule looks like:
```
{ghc,f} -> [BuildModule,hi,ho,dyn_hi,dyn_o]
```
In my proposed #416 it would be
```
{ghc,f} -> BuildModule -> {hi,ho,dyn_hi,dyn_o}
```
But the proper dependency graph has separate nodes for static and dynamic:
```
{ghc,f} -> HscOut -> {StaticBuild,DynamicBuild}
StaticBuild->{hi,ho}
DynamicBuild->{dyn_hi,dyn_o}
```
I'm not sure how GHC stores the pipeline phases, but presumably whatever method it uses to share HscOut between static and dynamic builds can be replicated in Shake. But in general it's not replicable, so Shake needs some kind of meta-build-rule that can look at which of StaticBuild or DynamicBuild are needed and run ghc with the respective ``-static``, ``-dynamic``, or ``-dynamic-too`` flag.

**Comment by ezyang** [comment]
I don't entirely understand your dependency graph syntax.

Pipeline intermediate phases are not storable because GHC doesn't support serializing it to disk. So if we black box GHC we can't replicate it in Shake.

**Comment by Mathnerd314** [comment]
{a,b} -> c is shorthand for a -> c, b -> c; is there anything else incomprehensible?

Right, GHC as a black box doesn't let you store phases. But Shake could still model the phases, and use that to determine which arguments to call GHC with.

**Comment by ezyang** [comment]
OK, I think what i implemented does have the dependency graph `{ghc,f} -> BuildModule -> {hi,ho,dyn_hi,dyn_o}`. In fact, in ghc-shake right now, you can't even depend on an hi file (so it's really only `{ghc,f} -> BuildModule`.

I do agree that the proper dependency graph is "better", but I am not sure if it's backwards compatible with the interface that `ghc --make` provides. In particular, `-dynamic` dumps the dynamic object files not in the name `.dyn_o`, but `.o` (it is dynamic-too which actually manges the renaming.)

**Comment by Mathnerd314** [comment]
Yes, I believe yours is just ``BuildModule``. I've updated it with actual dot diagrams.

**Comment by ndmitchell** [comment]
So what are the advantage of wiring up the `hi` and `o` files? Using a hypothetical `provide` function (which I think #416 introduces) would you be able to `need [file.hi]` and have it built? How would Shake know which rule might `provide` that without running them all? If so then the nodes seem fundamentally different to the other nodes on the graph. And if that's right, what properties do those nodes have? How do they interact? What are the benefits from having them in the graph?

**Comment by ezyang** [comment]
For what it's worth, I think it's good that Shake is target-oriented as opposed to source-oriented, because (to me at least) it feels like, within a build system, you more frequently want to answer the question "how do I build this" as opposed to "here is a source file please build it"; the latter generally only occurs at the user-level. Target-orientation versus source-orientation have complimentary weaknesses: in a target-oriented system, you have to do a bit of work to figure out what the result of building a source file is (that's the topic of this ticket: GHC is source-oriented and so some work must be done to offer that interface from a target-oriented system); but in a source-oriented system, you have to do a bit of work to figure out what source actually produces a given output entity.

**Comment by ndmitchell** [comment]
The development version of Shake now contains [Development.Shake.Forward](https://github.com/ndmitchell/shake/blob/master/src/Development/Shake/Forward.hs), which is a source-orientated build system. I suspect that for large build systems (e.g. > 200 lines of build code) target-oriented is probably easier to write.

**Comment by ndmitchell** [comment]
Rereading this quite a few times my best guess is this is _not_ an instance of the "unknown result" pattern, and would not benefit from a forward/source-directed build system. The fundamental problem is that you want a version of `&%>` which instead of requiring all files to be produced, sometimes produces different numbers of them. I'm hoping to spend an hour figuring this out tonight...

**Comment by ndmitchell** [comment]
OK, I sat down and figured out what I'd write. I then went looking to see what you actually did write. They are clearly the same idea. So, to summarise what I "recommend" (which may indicate where I differ from you and didn't realise it):

* You introduce a new type of rule. The key is a module name, the value is a tuple of Maybe FileA's, one for each of the hi/o/hi_dyn/o_dyn files.
* You put the dyn/no-dyn flag in the ShakeOptions in shakeExtra.
* The `storedValue` for this type queries all the values.
* The `equalValue` for this type checks only the ones that are relevant, e.g. no-dyn doesn't check hi_dyn.
* The build rule calls ghc with either dyn/no-dyn and builds the relevant pieces. For anything it doesn't build it fills in Nothing, regardless of whether it is present or not.
* You add general rules for each of .hi/.o/.hi_dyn/.o_dyn 

I think that accurately models the build system. Do you agree? I also think the solution wasn't at all obvious, so documenting "the pattern" is worthwhile (I think in generality you have a set of modes, and build a subset at any one time). And once we've got all that, then we can think if the implementation should be easier, e.g. composing files to make a "multi-file" rule. The `&%>` seems a very specific instance of a multi-file rule (all required), but clearly it's not the only one.

**Comment by ezyang** [comment]
I agree with this characterization. I don't think it's a very good pattern to document, though, because you can't actually do it with Shake as released today (FileA is not exported so I can't build it up this way.)

**Comment by ndmitchell** [comment]
I'm making the assumption that either FileA will be exposed in the short term (I wanted to understand this example first, and now I think I do). I'm also hopeful that we can make it a bit less manual labour than it currently is. I hope this remains the advanced aspects of Shake though, not for normal users.
